server/coordinator/procedure/apply.go

```diff
@@ -0,0 +1,172 @@
+	}
+)
+
+type ApplyCallbackRequest struct {
+	ctx context.Context
+
+	nodeName                   string
+	shardsNeedToReopen         []cluster.ShardInfo
+	shardsNeedToCloseAndReopen []cluster.ShardInfo
+	shardsNeedToClose          []cluster.ShardInfo
+
+	dispatch eventdispatch.Dispatch
+}
+
+// ApplyProcedure used for apply the latest cluster topology to CeresDB node when metadata version is not equal to CeresDB node version.
+type ApplyProcedure struct {
+	id       uint64
+	fsm      *fsm.FSM
+	dispatch eventdispatch.Dispatch
+
+	lock  sync.RWMutex
+	state State
+
+	nodeName                   string
+	shardsNeedToReopen         []cluster.ShardInfo
+	shardsNeedToCloseAndReopen []cluster.ShardInfo
+	shardsNeedToClose          []cluster.ShardInfo
+}
+
+func NewApplyProcedure(dispatch eventdispatch.Dispatch, id uint64, nodeName string, shardsNeedToReopen []cluster.ShardInfo, shardsNeedToCloseAndReopen []cluster.ShardInfo, shardNeedToClose []cluster.ShardInfo) Procedure {
+	return &ApplyProcedure{
+		id:                         id,
+		fsm:                        fsm.NewFSM(stateApplyBegin, applyEvents, applyCallbacks),
+		nodeName:                   nodeName,
+		shardsNeedToReopen:         shardsNeedToReopen,
+		shardsNeedToClose:          shardNeedToClose,
+		shardsNeedToCloseAndReopen: shardsNeedToCloseAndReopen,
+		dispatch:                   dispatch,
+	}
+}
+
+func (p *ApplyProcedure) ID() uint64 {
+	return p.id
+}
+
+func (p *ApplyProcedure) Typ() Typ {
+	return Apply
+}
+
+func (p *ApplyProcedure) Start(ctx context.Context) error {
+	log.Info("apply procedure start", zap.Uint64("procedureID", p.id))
+
+	p.updateStateWithLock(StateRunning)
+
+	applyCallbackRequest := &ApplyCallbackRequest{
+		ctx:                        ctx,
+		nodeName:                   p.nodeName,
+		shardsNeedToReopen:         p.shardsNeedToReopen,
+		shardsNeedToCloseAndReopen: p.shardsNeedToCloseAndReopen,
+		shardsNeedToClose:          p.shardsNeedToClose,
+		dispatch:                   p.dispatch,
+	}
+
+	if err := p.fsm.Event(eventApplyPrepare, applyCallbackRequest); err != nil {
+		err1 := p.fsm.Event(eventApplyFailed, applyCallbackRequest)
+		p.updateStateWithLock(StateFailed)
+		if err1 != nil {
+			err = errors.WithMessagef(err, "apply procedure start, fail to send eventApplyFailed err:%v", err1)
+		}
+		return errors.WithMessage(err, "apply procedure start")
+	}
+
+	if err := p.fsm.Event(eventApplySuccess, applyCallbackRequest); err != nil {
+		return errors.WithMessage(err, "apply procedure start")
+	}
+
+	p.updateStateWithLock(StateFinished)
+	return nil
+}
+
+func (p *ApplyProcedure) Cancel(_ context.Context) error {
+	p.updateStateWithLock(StateCancelled)
+	return nil
+}
+
+func (p *ApplyProcedure) State() State {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	return p.state
+}
+
+func (p *ApplyProcedure) updateStateWithLock(state State) {
+	p.lock.Lock()
+	p.state = state
+	p.lock.Unlock()
+}
+
+func applyPrepareCallback(event *fsm.Event) {
```

Comment:
`applyPrepareCallback` needs to be divided into several stages, such as `Close old shard`, `Open new shard` and so on.